### PR TITLE
Make source IP address configurable

### PR DESCRIPTION
**Pull request by AlexHockey**
This PR adds an option to libmemcached, `--SOURCE_ADDRESS` that allows the user to specify which IP address should be used for a given `memcached_st` instance. This change is needed as part of https://github.com/Metaswitch/cpp-common/pull/750.

There are seemingly lots of odd changes in this PR. This is because the parser code was all autogenerated a long time ago, and the checked in, because the parser toolchain that libmemcached used changed is behavior, and libmemcached did not keep up. I re-generated this code with the old toolchain and checked in the new autogenerated code, but it's probably best to ignore those files (parser.h, parser.cc, scanner.h and scanner.cc).

### CODE REVIEW

**Comment by graemerobertson** [comment]
Good God...

File: libmemcached-1.0/struct/memcached.h [hunk 1]
```diff
@@ -124,6 +124,7 @@ struct memcached_st {
     uint32_t max_pool_size;
     int32_t version; // This is used by pool and others to determine if the memcached_st is out of date.
     struct memcached_array_st *filename;
+    struct memcached_array_st *source_address;
```
Comment (graemerobertson) [inline libmemcached-1.0/struct/memcached.h]: I don't know what a `memcached_array_st` is and I'm pretty certain I don't want to know either.

Maybe an array of characters that looks a little bit like a string, but is _specific_ to memcached in some way. Maybe the string is terminated with `6d 65 6d 63 61 63 68 65 64`.

File: libmemcached/connect.cc [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*  vim:expandtab:shiftwidth=2:tabstop=2:smarttab:
- *
```
Comment (graemerobertson) [inline libmemcached/connect.cc]: +1

File: libmemcached/csl/parser.cc [hunk 3]
```diff
@@ -1797,379 +1807,379 @@ YYSTYPE yylval;
...
             (yyval.behavior)= MEMCACHED_BEHAVIOR_CONNECT_TIMEOUT;
           }
     break;
 
-  case 29:
+  case 30:
 
 /* Line 1806 of yacc.c  */
-#line 354 "libmemcached/csl/parser.yy"
+#line 359 "libmemcached/csl/parser.yy"
```
Comment (graemerobertson) [inline libmemcached/csl/parser.cc]: Is this right?

File: libmemcached/csl/parser.cc [hunk 4]
```diff
@@ -705,11 +706,11 @@ static const yytype_uint8 yystos[] =
       12,    17,    18,    19,    20,    21,    22,    23,    24,    25,
       28,    29,    30,    31,    32,    33,    34,    35,    36,    37,
       38,    39,    40,    41,    42,    43,    44,    45,    46,    47,
-      48,    49,    77,    78,    79,    80,    81,    82,    75,    72,
-      73,    86,    70,    71,    86,    59,    60,    61,    87,    50,
-      51,    52,    53,    54,    55,    56,    57,    58,    85,    86,
-      67,    67,     0,    75,    67,    86,    68,    83,    83,    69,
-      84,    64,    78,    84,    84,    85
```
Comment (graemerobertson) [inline libmemcached/csl/parser.cc]: Did this table of random numbers require re-factoring?

File: libmemcached/csl/scanner.cc [hunk 5]
```diff
@@ -463,76 +463,77 @@ static void yy_fatal_error (yyconst char msg[] ,yyscan_t yyscanner );
...
 	};
-static yyconst flex_int16_t yy_accept[534] =
+static yyconst flex_int16_t yy_accept[547] =
     {   0,
-        0,    0,   67,   65,    5,    5,    1,   65,   65,   65,
-        2,   65,   65,   65,   65,   65,   65,   65,   65,   65,
-       65,   65,   65,   65,   65,   65,    0,   64,    0,   49,
+        0,    0,   68,   66,    5,    5,    1,   66,   66,   66,
+        2,   66,   66,   66,   66,   66,   66,   66,   66,   66,
+       66,   66,   66,   66,   66,   66,    0,   65,    0,   50,
```
Comment (graemerobertson) [inline libmemcached/csl/scanner.cc]: YES! Finally someone has fixed this up

File: libmemcached/csl/scanner.l [hunk 6]
```diff
@@ -150,6 +150,7 @@
 "--NAMESPACE="	       		        { yyextra->begin= yytext; return yyextra->previous_token= NAMESPACE; }
 
 "--FETCH-VERSION"	       		        { yyextra->begin= yytext; return yyextra->previous_token= FETCH_VERSION; }
+"--SOURCE-ADDRESS="			{ yyextra->begin= yytext; return yyextra->previous_token= SOURCE_ADDRESS; }
```
Comment (graemerobertson) [inline libmemcached/csl/scanner.l]: Indentation looks a bit off here

File: libmemcached/csl/scanner.cc (hunk 5 above)
Comment (AlexHockey) [reply libmemcached/csl/scanner.cc]: I know - it's so nice to finally be tackling this annoying technical debt. Finally now our velocity can start increasing!

**Pull request merged**